### PR TITLE
Refactor and improve camera geometry and image cleaning

### PR DESCRIPTION
**Pull request by kosack**
Improvements
---------------

- based on speed-ups in issue #346
- moved `CameraGeometry` from `io` to `instrument`, and replaced the duplicate functionality in instrument
- sped up `CameraGeometry.guess()`, which now uses a cache to no longer recompute the same geometry multiple times
- neighbor lists are now lazily evaluated, for speed
- now provides a neighbor_matrix in addition to the neighbor list, for faster algorithms
- speed up `image.tailcuts_clean` and `image.dilate` by using `CameraGeometry.neighbor_matrix` (speed is faster than standard c implementation!) 

API changes
-------------
-  `image.tailcuts_clean` no longer takes a `pedvars` argument, which was redundant in most cases 
-  all references to `ctapipe.io.CameraGeometry` should now be c `ctapipe.instrument.CameraGeometry`

### CODE REVIEW

**Comment by kosack** [comment]
@watsonjj  do you see any problems with this, or rather, does it break anything major in your code?

File: ctapipe/instrument/camera.py [hunk 1]
```diff
@@ -33,19 +33,35 @@
 }
 
 
+_geometry_cache = {}
```
Comment (watsonjj) [inline ctapipe/instrument/camera.py]: Is this an erroneous declaration?

Comment (kosack) [inline ctapipe/instrument/camera.py]: no, it's used later in the guess() method. It's a class attribute to keep the list of previously constructed geometries, which are keyed by an integer hash that combines the x,y vectors and the focal length (which I think is enough to make them unique).  Originally, I had used only the number of pixels, but if we use the hexagonal-to-cartesian camera transforms, the number of pixels stays the same but the pixel ordering changes, so it caused some problems.

Comment (kosack) [inline ctapipe/instrument/camera.py]: I'll add a comment to that line

Comment (watsonjj) [inline ctapipe/instrument/camera.py]: But there are 2 `_geometry_cache = {}`, one global on line 36, and one as an attribute of CameraGeometry on line 60, are both required?

Comment (kosack) [inline ctapipe/instrument/camera.py]: ahh, yes that is a mistake. The global one is not needed. I'll remove it

**Comment by kosack** [comment]
if there's no major issue, I will merge this soon - I'd like to work on fixing a few other bugs/features that depend on this .

**Comment by watsonjj** [comment]
I'm happy for this to be merged. I should implement the use of the neighbour_matrix in the charge extraction algorithms, correct? 

Is the old neighbour list going to be deprecated eventually? I see that neighbour_matrix relies on that list for its creation, is that still the best way for it to be created?

**Comment by kosack** [comment]
> I'm happy for this to be merged. I should implement the use of the neighbour_matrix in the charge extraction algorithms, correct?

Yes that would be great. I already replaced a call in   `dl1.py` that calculated the neighbors again with a call to `CameraGeometry.neighbors`, and that simple change sped up the code (at least by a few seconds of runtime), now that CameraGeometry is cached. It still uses the neighbor list and not the matrix though, so changing the integrators to use it would improve the speed even more.

>Is the old neighbour list going to be deprecated eventually? I see that neighbour_matrix relies on that list for its creation, is that still the best way for it to be created?

I left both in there for now to not break everything, but we could eventually deprecate the neighbor list if it is not used. It could be useful in some cases.  It does not need to be created first in order to make the neighbor_matrix, I just did it that way for simplicity (the matrix could easily be generated directly).  For now, I think we can leave both, and recommend the matrix for fast calculations.  I think it may be nicer to have the list in case we want to write it to a  human-readable file though.

File: ctapipe/calib/camera/dl1.py [hunk 2]
```diff
@@ -176,7 +176,7 @@ def check_dl0_exists(self, event, telid):
                 self._dl0_empty_warn = True
             return False
 
-    def get_neighbours(self, event, telid):
+    def get_geometry(self, event, telid):
         """
         Obtain the neighbouring pixels for this telescope.
```
Comment (MaxNoe) [inline ctapipe/calib/camera/dl1.py]: This is now lying documentation

Comment (watsonjj) [inline ctapipe/calib/camera/dl1.py]: I will fix that in my next PR

**Comment by watsonjj** [comment]
@kosack Is it possible to change CameraGeometry to a Component? This would allow the radius size for the pixel neighbours to be changed, but I'm not sure if it makes sense to change CameraGeometry like that. Previously this could be changed by the radius traitlet in dl1.py.

**Pull request merged**